File: reverse/game1/3-readable/tjge/SpriteDef.java
// 派生文件 · 非权威 · 逻辑逐字节等同 2-decompiled-cfr/tjge/d.java
// 标识符按 reverse/game1/3-readable/SYMBOLS.md 重命名，仅供阅读；任何不一致以 CFR 为准。
/*
 * Decompiled with CFR 0.152.
 *
 * Could not load the following classes:
 *  javax.microedition.lcdui.Graphics
 */
package tjge;

import java.io.InputStream;
import javax.microedition.lcdui.Graphics;
import tjge.GameMIDlet;
import tjge.GameScreen;       // 原 tjge.a
import tjge.SpriteAtlas;      // 原 tjge.i

public final class SpriteDef {
    // 静态：图集(SpriteAtlas)表，按图集 id 索引，容量 32，懒加载缓存。            (原 a)
    public static SpriteAtlas[] atlasTable = new SpriteAtlas[32];
    // 静态：各图集 id 的已加载标志表，容量 32；索引 0 在静态块中置 true。           (原 b)
    public static boolean[] atlasLoadedFlags = new boolean[32];
    // 静态：资源条目索引 → 图集 id 的映射表，容量 32。                          (原 c)
    public static int[] resourceToAtlasId = new int[32];

    private short poseCount;                 // pose（拼帧姿态）数量。                                    (原 o)
    private short[] poseModuleCounts;        // 每个 pose 的模块(分块)数量。                              (原 p)
    private short[][] poseModuleFrames;      // 每个 pose 各模块在图集中的帧索引(读取时负值+256还原)。      (原 q)
    public byte[][] poseModuleOffsetX;       // 每个 pose 各模块的 x 偏移。                               (原 d)
    public byte[][] poseModuleOffsetY;       // 每个 pose 各模块的 y 偏移。                               (原 e)
    private short sequenceCount;             // sequence（动画序列）数量。                                (原 r)
    private short[] sequenceFrameCounts;     // 每个 sequence 的帧数。                                    (原 s)
    private short[][] sequencePoseIndices;   // 每个 sequence 的逐帧 pose 索引序列(读取时负值+256还原)。   (原 t)
    public byte[] collisionBoxX;             // 每个 sequence 的碰撞箱 x。                                (原 f)
    public byte[] collisionBoxY;             // 每个 sequence 的碰撞箱 y。                                (原 g)
    public byte[] collisionBoxWidth;         // 每个 sequence 的碰撞箱宽 w。                              (原 h)
    public byte[] collisionBoxHeight;        // 每个 sequence 的碰撞箱高 h。                              (原 i)
    public short boundsX;                    // 整体边界 x。                                              (原 j)
    public short boundsY;                    // 整体边界 y。                                              (原 k)
    public short boundsWidth;                // 整体边界宽。                                              (原 l)
    public short boundsHeight;               // 整体边界高。                                              (原 m)
    public short atlasId;                    // 图集 id（临时字段：解析时读入并关联图集，末尾被清零）。      (原 n)
    private SpriteAtlas atlas;               // 本精灵关联的图集，绘制时调用其拼块绘制。                    (原 u)

    // 私有构造函数，仅供静态工厂 loadFromBin(int) 内部创建实例。
    private SpriteDef() {
    }

    // 返回 sequence（动画序列）数量。                                          (原 a()S)
    public final short getSequenceCount() {
        return this.sequenceCount;
    }

    // 返回第 sequenceIndex 个 sequence 的帧数。                                (原 a(I)S)
    public final short getSequenceFrameCount(int sequenceIndex) {
        return this.sequenceFrameCounts[sequenceIndex];
    }

    // 按 sequence 索引(高2位为水平/垂直翻转标志)与帧索引，把对应 pose 的各模块帧逐块拼到 Graphics 上绘制。
    // (原 a(Graphics;IIIIII)V)
    public final void paintSequenceFrame(Graphics graphics, int x, int y, int sequenceIndexFlags, int frameIndex, int n5, int orientation) {
        int flipHorizontal = sequenceIndexFlags & Integer.MIN_VALUE;
        int flipVertical = sequenceIndexFlags & 0x40000000;
        short pose = this.sequencePoseIndices[sequenceIndexFlags &= 0xFFFFFF][frameIndex];
        int moduleCount = this.poseModuleCounts[pose];
        graphics.setClip(0, 0, 176, 208);
        int module = 0;
        while (module < moduleCount) {
            int drawY;
            int drawX;
            short moduleFrame = this.poseModuleFrames[pose][module];
            byte offsetX = this.poseModuleOffsetX[pose][module];
            if (flipHorizontal != 0) {
                offsetX = (byte)(-offsetX);
            }
            byte offsetY = this.poseModuleOffsetY[pose][module];
            if (flipVertical != 0) {
                offsetY = (byte)(-offsetY);
            }
            byte offsetXFinal = offsetX;
            byte offsetYFinal = offsetY;
            int n13 = offsetYFinal + this.atlas.heights[moduleFrame];
            if (orientation == 270) {
                drawX = x - n13;
                drawY = y + offsetXFinal;
            } else {
                drawX = x + offsetX;
                drawY = y + offsetY;
            }
            this.atlas.drawSprite(graphics, drawX, drawY, moduleFrame | flipHorizontal | flipVertical, n5, orientation);
            ++module;
        }
    }

    // 释放本精灵持有的所有 pose/sequence/碰撞箱数组与图集引用。                  (原 b()V)
    public final void dispose() {
        this.collisionBoxX = null;
        this.collisionBoxY = null;
        this.collisionBoxWidth = null;
        this.collisionBoxHeight = null;
        int n = 0;
        while (n < this.sequencePoseIndices.length) {
            this.sequencePoseIndices[n] = null;
            ++n;
        }
        this.sequencePoseIndices = null;
        int n2 = 0;
        while (n2 < this.poseModuleFrames.length) {
            this.poseModuleFrames[n2] = null;
            this.poseModuleOffsetX[n2] = null;
            this.poseModuleOffsetY[n2] = null;
            ++n2;
        }
        this.poseModuleFrames = null;
        this.poseModuleOffsetX = null;
        this.poseModuleOffsetY = null;
        this.poseModuleCounts = null;
        this.sequenceFrameCounts = null;
        this.atlas = null;
    }

    // 静态：依据 GameScreen.isScrollLevel(int) 刷新各资源索引对应图集的已加载标志表。   (原 c()V)
    public static final void refreshLoadedAtlasFlags() {
        int n = 1;
        while (n < 32) {
            tjge.SpriteDef.atlasLoadedFlags[tjge.SpriteDef.resourceToAtlasId[n]] = tjge.GameScreen.isScrollLevel(n);
            ++n;
        }
    }

    // 静态工厂：从 /res/a.bin 第 entryIndex 条目解析出 SpriteDef（pose 拼帧 + sequence 序列 + 碰撞箱），失败返回 null。
    // (原 b(I)Ltjge/d;)
    public static final SpriteDef loadFromBin(int entryIndex) {
        SpriteDef spriteDef = new SpriteDef();
        try {
            int pose;
            InputStream inputStream = GameMIDlet.openArchiveEntryStream("/res/a.bin", entryIndex);
            spriteDef.atlasId = GameMIDlet.readU16Le(inputStream);
            if (atlasTable[spriteDef.atlasId] == null) {
                tjge.SpriteDef.atlasTable[spriteDef.atlasId] = tjge.SpriteAtlas.load(spriteDef.atlasId);
            }
            tjge.SpriteDef.atlasLoadedFlags[spriteDef.atlasId] = true;
            spriteDef.atlas = atlasTable[spriteDef.atlasId];
            tjge.SpriteDef.resourceToAtlasId[entryIndex] = spriteDef.atlasId;
            spriteDef.poseCount = GameMIDlet.readU16Le(inputStream);
            spriteDef.poseModuleCounts = new short[spriteDef.poseCount];
            spriteDef.poseModuleFrames = new short[spriteDef.poseCount][];
            spriteDef.poseModuleOffsetX = new byte[spriteDef.poseCount][];
            spriteDef.poseModuleOffsetY = new byte[spriteDef.poseCount][];
            int poseIndex = 0;
            while (poseIndex < spriteDef.poseCount) {
                spriteDef.poseModuleCounts[poseIndex] = GameMIDlet.readU16Le(inputStream);
                spriteDef.poseModuleFrames[poseIndex] = new short[spriteDef.poseModuleCounts[poseIndex]];
                spriteDef.poseModuleOffsetX[poseIndex] = new byte[spriteDef.poseModuleCounts[poseIndex]];
                spriteDef.poseModuleOffsetY[poseIndex] = new byte[spriteDef.poseModuleCounts[poseIndex]];
                pose = 0;
                while (pose < spriteDef.poseModuleCounts[poseIndex]) {
                    spriteDef.poseModuleFrames[poseIndex][pose] = GameMIDlet.readByte(inputStream);
                    if (spriteDef.poseModuleFrames[poseIndex][pose] < 0) {
                        short[] frames = spriteDef.poseModuleFrames[poseIndex];
                        int idx = pose;
                        frames[idx] = (short)(frames[idx] + 256);
                    }
                    spriteDef.poseModuleOffsetX[poseIndex][pose] = GameMIDlet.readByte(inputStream);
                    spriteDef.poseModuleOffsetY[poseIndex][pose] = GameMIDlet.readByte(inputStream);
                    ++pose;
                }
                ++poseIndex;
            }
            inputStream.read();
            inputStream.read();
            inputStream.read();
            inputStream.read();
            spriteDef.sequenceCount = GameMIDlet.readU16Le(inputStream);
            spriteDef.collisionBoxX = new byte[spriteDef.sequenceCount];
            spriteDef.collisionBoxY = new byte[spriteDef.sequenceCount];
            spriteDef.collisionBoxWidth = new byte[spriteDef.sequenceCount];
            spriteDef.collisionBoxHeight = new byte[spriteDef.sequenceCount];
            spriteDef.sequenceFrameCounts = new short[spriteDef.sequenceCount];
            spriteDef.sequencePoseIndices = new short[spriteDef.sequenceCount][];
            pose = 0;
            while (pose < spriteDef.sequenceCount) {
                spriteDef.collisionBoxX[pose] = GameMIDlet.readByte(inputStream);
                spriteDef.collisionBoxY[pose] = GameMIDlet.readByte(inputStream);
                spriteDef.collisionBoxWidth[pose] = GameMIDlet.readByte(inputStream);
                spriteDef.collisionBoxHeight[pose] = GameMIDlet.readByte(inputStream);
                spriteDef.sequenceFrameCounts[pose] = GameMIDlet.readU16Le(inputStream);
                spriteDef.sequencePoseIndices[pose] = new short[spriteDef.sequenceFrameCounts[pose]];
                int frame = 0;
                while (frame < spriteDef.sequenceFrameCounts[pose]) {
                    spriteDef.sequencePoseIndices[pose][frame] = GameMIDlet.readByte(inputStream);
                    if (spriteDef.sequencePoseIndices[pose][frame] < 0) {
                        short[] poses = spriteDef.sequencePoseIndices[pose];
                        int idx = frame;
                        poses[idx] = (short)(poses[idx] + 256);
                    }
                    ++frame;
                }
                ++pose;
            }
            spriteDef.boundsX = GameMIDlet.readU16Le(inputStream);
            spriteDef.boundsY = GameMIDlet.readU16Le(inputStream);
            spriteDef.boundsWidth = GameMIDlet.readU16Le(inputStream);
            spriteDef.boundsHeight = GameMIDlet.readU16Le(inputStream);
            spriteDef.atlasId = 0;
            inputStream.close();
        }
        catch (Exception exception) {
            return null;
        }
        return spriteDef;
    }

    static {
        tjge.SpriteDef.atlasLoadedFlags[0] = true;
    }
}

File: packages/game1/src/SpriteDef.ts
/**
 * 游戏1《红魔特种兵》精灵帧定义/拼帧绘制。
 * 逐行移植自 reverse/game1/2-decompiled-cfr/tjge/d.java（CFR 权威源）。
 * 移植规约见 docs/05-移植规约.md；方法名映射见 reverse/game1/porting-naming/porting-contract.json。
 *
 * 职责：解析 a.bin（两级动画：pose 拼帧 + sequence 序列 + 碰撞箱），
 *       提供拼帧绘制 a_GIIIIII（按 pose 把模块帧逐块拼到 Graphics）。
 *
 * 字段别名：
 *   静态 a=精灵图集(i)表, b=已加载标志表, c=资源索引→图集id映射
 *   o=pose 数量, p=每个 pose 的模块数, q=每个 pose 各模块的帧索引(0..255)
 *   d=每个 pose 各模块的 x 偏移(byte), e=每个 pose 各模块的 y 偏移(byte)
 *   r=sequence 数量, s=每个 sequence 的帧数, t=每个 sequence 的 pose 索引序列
 *   f/g/h/i=每个 sequence 的碰撞箱 x/y/w/h, j/k/l/m=整体边界, n=图集id(临时)
 *   u=本精灵关联的图集(i)
 *
 * 方法映射：a()→a_, a(int)→a_I, a(Graphics,6×int)→a_GIIIIII,
 *           b()→b_, c()→c_(static), b(int)→b_I(static), 构造→constructor
 */
import { Graphics } from "@red-devil/j2me-shim";
import { GameMIDlet } from "./GameMIDlet.ts";
import { GameScreen } from "./GameScreen.ts";
import { SpriteAtlas } from "./SpriteAtlas.ts";
import { MIRROR_FLAG, FLIP_VERTICAL_BIT, SEQUENCE_MASK } from "./constants.ts";

/**
 * 精灵帧/动画定义（原 CFR `tjge.d`，基准文件 reverse/game1/2-decompiled-cfr/tjge/d.java）。
 *
 * 角色：游戏1《红魔特种兵》的两级动画资源描述符。解析归档 a.bin 的单条目，得到一套
 * 「pose（拼帧姿态）+ sequence（动画序列）+ 碰撞箱」数据，是所有演员（Player/Enemy/Boss/
 * Projectile/Effect/Pickup）绘制与碰撞的底层数据来源。每个精灵类型对应一个 SpriteDef 实例，
 * 由 LevelLoader 引用计数管理、由 GameScreen.createActor 传给演员；演员只持有 pose/sequence
 * 索引，真正的逐块拼帧绘制委托给本类的 {@link paintSequenceFrame}。
 *
 * 数据模型（两级动画）：
 *   - pose：一个静态拼帧姿态，由若干「模块」组成，每个模块 = 图集小图帧索引 + (x,y) 偏移；
 *   - sequence：一段动画，由一串 pose 索引构成（逐帧切换 pose），并附带该动画的碰撞箱。
 *
 * 关键字段（混淆名→语义，详见 reverse/game1/3-readable/SYMBOLS.md「SpriteDef」节）：
 *   - 静态 atlasTable=图集(SpriteAtlas)缓存表[32]，atlasLoadedFlags=各图集已加载标志[32]，
 *     resourceToAtlasId=资源条目索引→图集 id 映射[32]；
 *   - poseCount/poseModuleCounts/poseModuleFrames=pose 数量/每 pose 模块数/每模块图集帧索引；
 *   - poseModuleOffsetX/poseModuleOffsetY=每模块的有符号 x/y 偏移（绘制时按翻转标志取负）；
 *   - sequenceCount/sequenceFrameCounts/sequencePoseIndices=动画数量/每动画帧数/每帧的 pose 索引；
 *   - collisionBoxX/Y/Width/Height=每个 sequence 的碰撞箱（有符号 byte）；
 *   - boundsX/Y/Width/Height=精灵整体边界；atlasId=临时图集 id；atlas=本精灵关联的图集对象。
 *
 * 协作者：SpriteAtlas（提供 drawSprite 贴图与 heights 表）、GameMIDlet（提供 .bin 流定位与小端读取）、
 * GameScreen（提供 isScrollLevel 判定，供 refreshLoadedAtlasFlags 使用）。
 */
export class SpriteDef {
  static atlasTable: (SpriteAtlas | null)[] = new Array<SpriteAtlas | null>(32).fill(null);
  static atlasLoadedFlags: boolean[] = new Array<boolean>(32).fill(false);
  static resourceToAtlasId: Int32Array = new Int32Array(32);

  private poseCount: number = 0; // short
  private poseModuleCounts!: Int16Array; // short[]
  private poseModuleFrames!: Int16Array[]; // short[][]
  poseModuleOffsetX!: Int8Array[]; // byte[][]
  poseModuleOffsetY!: Int8Array[]; // byte[][]
  private sequenceCount: number = 0; // short
  private sequenceFrameCounts!: Int16Array; // short[]
  private sequencePoseIndices!: Int16Array[]; // short[][]
  collisionBoxX!: Int8Array; // byte[]
  collisionBoxY!: Int8Array; // byte[]
  collisionBoxWidth!: Int8Array; // byte[]
  collisionBoxHeight!: Int8Array; // byte[]
  boundsX: number = 0; // short
  boundsY: number = 0; // short
  boundsWidth: number = 0; // short
  boundsHeight: number = 0; // short
  atlasId: number = 0; // short
  private atlas!: SpriteAtlas;

  /** d.<init>()：私有构造函数，仅供静态工厂 {@link loadFromBin} 内部创建实例（外部应走工厂方法）。 */
  private constructor() {
  }

  /** d.a()：返回 sequence 数量 r。 */
  getSequenceCount(): number {
    return this.sequenceCount;
  }

  /** d.a(int)：返回第 n 个 sequence 的帧数 s[n]。 */
  getSequenceFrameCount(sequenceId: number): number {
    return this.sequenceFrameCounts[sequenceId];
  }

  /**
   * d.a(Graphics,int,int,int,int,int,int)：按 sequence/帧拼帧绘制。
   * x/y=坐标, sequence=序列索引(高2位为水平/垂直翻转标志), frame=帧索引, drawParamA/rotation=透传给图集绘制(缩放/旋转, rotation===270 时交换)。
   */
  paintSequenceFrame(graphics: Graphics, x: number, y: number, sequence: number, frame: number, drawParamA: number, rotation: number): void {
    const mirrorFlag = sequence & MIRROR_FLAG; // Integer.MIN_VALUE (0x80000000)
    const flipFlag = sequence & FLIP_VERTICAL_BIT;
    const pose = this.sequencePoseIndices[(sequence &= SEQUENCE_MASK)][frame];
    const moduleCount = this.poseModuleCounts[pose];
    graphics.setClip(0, 0, 176, 208);
    let i = 0;
    while (i < moduleCount) {
      let drawY: number;
      let drawX: number;
      const module = this.poseModuleFrames[pose][i];
      let offsetX = this.poseModuleOffsetX[pose][i];
      if (mirrorFlag !== 0) {
        offsetX = ((-offsetX) << 24) >> 24; // (byte)(-offsetX)：Java byte 取负带 i2b 截断，-128 取负仍为 -128
      }
      let offsetY = this.poseModuleOffsetY[pose][i];
      if (flipFlag !== 0) {
        offsetY = ((-offsetY) << 24) >> 24; // (byte)(-offsetY)
      }
      const shiftX = offsetX;
      const shiftY = offsetY;
      const rotatedHeight = shiftY + this.atlas.heights[module];
      if (rotation === 270) {
        drawX = x - rotatedHeight;
        drawY = y + shiftX;
      } else {
        drawX = x + offsetX;
        drawY = y + offsetY;
      }
      this.atlas.drawSprite(graphics, drawX, drawY, module | mirrorFlag | flipFlag, drawParamA, rotation);
      ++i;
    }
  }

  /** d.b()：释放本精灵持有的所有数组引用。 */
  dispose(): void {
    this.collisionBoxX = null as unknown as Int8Array;
    this.collisionBoxY = null as unknown as Int8Array;
    this.collisionBoxWidth = null as unknown as Int8Array;
    this.collisionBoxHeight = null as unknown as Int8Array;
    let n = 0;
    while (n < this.sequencePoseIndices.length) {
      this.sequencePoseIndices[n] = null as unknown as Int16Array;
      ++n;
    }
    this.sequencePoseIndices = null as unknown as Int16Array[];
    let n2 = 0;
    while (n2 < this.poseModuleFrames.length) {
      this.poseModuleFrames[n2] = null as unknown as Int16Array;
      this.poseModuleOffsetX[n2] = null as unknown as Int8Array;
      this.poseModuleOffsetY[n2] = null as unknown as Int8Array;
      ++n2;
    }
    this.poseModuleFrames = null as unknown as Int16Array[];
    this.poseModuleOffsetX = null as unknown as Int8Array[];
    this.poseModuleOffsetY = null as unknown as Int8Array[];
    this.poseModuleCounts = null as unknown as Int16Array;
    this.sequenceFrameCounts = null as unknown as Int16Array;
    this.atlas = null as unknown as SpriteAtlas;
  }

  /** d.c()：刷新已加载图集表 b 的有效性（依据 a.d_I）。 */
  static refreshLoadedAtlasFlags(): void {
    let n = 1;
    while (n < 32) {
      SpriteDef.atlasLoadedFlags[SpriteDef.resourceToAtlasId[n]] = GameScreen.isScrollLevel(n);
      ++n;
    }
  }

  /**
   * d.b(int)：从 /res/a.bin 第 n 条目解析精灵帧定义。
   * 解析失败（异常）返回 null（保持原版吞异常返回 null 语义）。
   */
  static loadFromBin(entryIndex: number): SpriteDef | null {
    const spriteDef = new SpriteDef();
    try {
      let i: number;
      const inputStream = GameMIDlet.openArchiveEntryStream("/res/a.bin", entryIndex)!;
      spriteDef.atlasId = GameMIDlet.readU16Le(inputStream);
      if (SpriteDef.atlasTable[spriteDef.atlasId] == null) {
        SpriteDef.atlasTable[spriteDef.atlasId] = SpriteAtlas.load(spriteDef.atlasId);
      }
      SpriteDef.atlasLoadedFlags[spriteDef.atlasId] = true;
      spriteDef.atlas = SpriteDef.atlasTable[spriteDef.atlasId]!;
      SpriteDef.resourceToAtlasId[entryIndex] = spriteDef.atlasId;
      spriteDef.poseCount = GameMIDlet.readU16Le(inputStream);
      spriteDef.poseModuleCounts = new Int16Array(spriteDef.poseCount);
      spriteDef.poseModuleFrames = new Array<Int16Array>(spriteDef.poseCount);
      spriteDef.poseModuleOffsetX = new Array<Int8Array>(spriteDef.poseCount);
      spriteDef.poseModuleOffsetY = new Array<Int8Array>(spriteDef.poseCount);
      let poseIndex = 0;
      while (poseIndex < spriteDef.poseCount) {
        spriteDef.poseModuleCounts[poseIndex] = GameMIDlet.readU16Le(inputStream);
        spriteDef.poseModuleFrames[poseIndex] = new Int16Array(spriteDef.poseModuleCounts[poseIndex]);
        spriteDef.poseModuleOffsetX[poseIndex] = new Int8Array(spriteDef.poseModuleCounts[poseIndex]);
        spriteDef.poseModuleOffsetY[poseIndex] = new Int8Array(spriteDef.poseModuleCounts[poseIndex]);
        i = 0;
        while (i < spriteDef.poseModuleCounts[poseIndex]) {
          spriteDef.poseModuleFrames[poseIndex][i] = GameMIDlet.readByte(inputStream);
          if (spriteDef.poseModuleFrames[poseIndex][i] < 0) {
            const fixupArray = spriteDef.poseModuleFrames[poseIndex];
            const idx = i;
            fixupArray[idx] = ((fixupArray[idx] + 256) << 16) >> 16; // (short)
          }
          spriteDef.poseModuleOffsetX[poseIndex][i] = GameMIDlet.readByte(inputStream);
          spriteDef.poseModuleOffsetY[poseIndex][i] = GameMIDlet.readByte(inputStream);
          ++i;
        }
        ++poseIndex;
      }
      inputStream.read1();
      inputStream.read1();
      inputStream.read1();
      inputStream.read1();
      spriteDef.sequenceCount = GameMIDlet.readU16Le(inputStream);
      spriteDef.collisionBoxX = new Int8Array(spriteDef.sequenceCount);
      spriteDef.collisionBoxY = new Int8Array(spriteDef.sequenceCount);
      spriteDef.collisionBoxWidth = new Int8Array(spriteDef.sequenceCount);
      spriteDef.collisionBoxHeight = new Int8Array(spriteDef.sequenceCount);
      spriteDef.sequenceFrameCounts = new Int16Array(spriteDef.sequenceCount);
      spriteDef.sequencePoseIndices = new Array<Int16Array>(spriteDef.sequenceCount);
      i = 0;
      while (i < spriteDef.sequenceCount) {
        spriteDef.collisionBoxX[i] = GameMIDlet.readByte(inputStream);
        spriteDef.collisionBoxY[i] = GameMIDlet.readByte(inputStream);
        spriteDef.collisionBoxWidth[i] = GameMIDlet.readByte(inputStream);
        spriteDef.collisionBoxHeight[i] = GameMIDlet.readByte(inputStream);
        spriteDef.sequenceFrameCounts[i] = GameMIDlet.readU16Le(inputStream);
        spriteDef.sequencePoseIndices[i] = new Int16Array(spriteDef.sequenceFrameCounts[i]);
        let frameIndex = 0;
        while (frameIndex < spriteDef.sequenceFrameCounts[i]) {
          spriteDef.sequencePoseIndices[i][frameIndex] = GameMIDlet.readByte(inputStream);
          if (spriteDef.sequencePoseIndices[i][frameIndex] < 0) {
            const fixupArray = spriteDef.sequencePoseIndices[i];
            const idx = frameIndex;
            fixupArray[idx] = ((fixupArray[idx] + 256) << 16) >> 16; // (short)
          }
          ++frameIndex;
        }
        ++i;
      }
      spriteDef.boundsX = GameMIDlet.readU16Le(inputStream);
      spriteDef.boundsY = GameMIDlet.readU16Le(inputStream);
      spriteDef.boundsWidth = GameMIDlet.readU16Le(inputStream);
      spriteDef.boundsHeight = GameMIDlet.readU16Le(inputStream);
      spriteDef.atlasId = 0;
      inputStream.close();
    } catch (exception) {
      return null;
    }
    return spriteDef;
  }

  // 静态初始化块：原版 static { d.b[0] = true; }
  static {
    SpriteDef.atlasLoadedFlags[0] = true;
  }
}
